Changes made to bladeRF_MIMO_XCVR.m

danielwestofficial opened the pull request:
**Disclaimer: This is my very first pull request on GitHub!**

- Dynamic channel reconfiguration is enabled for synchronous TX and RX (lines of concern 241 through 258).
- Worked on the set.gain() function to allow the status to be read. (lines of concern 265 through 294)
- Setup defaults have been changed due to individual differences. (lines of concern 587 through 589)

I believe these are the only differences made to bladeRF_MIMO_XCVR.m, but there may be others please double check. 

I added the TestScript to ensure the bladeRF Ax4 2.0 is configured accurately. 
I also added a LoopBackTest script which is more specific to my FMCW application.

danielwestofficial committed:
Update bladeRF_MIMO_XCVR.m

Fixed the mistake in bladeRF_MIMO_XCVR.m in line 285. Where the fprintf prints bandwidth and not gain.

danielwestofficial committed:
Create TestScript

This script is used to ensure the hardware initialization is performed correctly.

danielwestofficial committed:
Rename TestScript to bladeRF_TestScript

danielwestofficial committed:
Create bladeRF_LoopbackTest

This script is specific for performing a loopback test on all 4 channels.

JoseAmador95 commented on bladeRF_MIMO_XCVR.m:
```diff
@@ -238,53 +238,84 @@
             end
+        end
 
+        % OLD VERION FROM Jose Amadaro "User Equipment...."
```
We can delete the commented code. It will still live in the git history.

danielwestofficial replied:
Deleted in most recent commit.

JoseAmador95 commented on bladeRF_MIMO_XCVR.m:
```diff
@@ -238,53 +238,84 @@
-            fprintf('Set %s%d bandwidth. Requested: [%d, %d], Actual: [%d, %d]\n', ...
-                    obj.direction, i, val, obj.gain)
-        end
+%            for i = 1:2
```
This chunk can be removed

danielwestofficial replied:
Removed in recent commit.

JoseAmador95 commented on bladeRF_MIMO_XCVR.m:
```diff
@@ -577,9 +608,9 @@
-            obj.samplerate = 1e6;
-            obj.frequency = 890e6;
-            obj.bandwidth = 1e6;
+            obj.samplerate = 2e6;
```
Can you please revert the default parameters?

danielwestofficial replied:
Default parameters reverted

JoseAmador95 commented on bladeRF_MIMO_XCVR.m:
```diff
@@ -597,7 +628,8 @@
             if strcmpi(dir,'RX') == true
                 for i = 1:2
-                    status = calllib('libbladeRF', 'bladerf_set_gain_mode', obj.bladerf.device, obj.module{i}, 'BLADERF_GAIN_DEFAULT');
+                    %status = calllib('libbladeRF', 'bladerf_set_gain_mode', obj.bladerf.device, obj.module{i}, 'BLADERF_GAIN_DEFAULT'); % Trying to fix RX gain assignment
```
This line can be deleted

danielwestofficial replied:
BLADERF_GAIN_MGC is used for manual gain control. Default AGC use's BLADERF_GAIN_DEFAULT. Changed but left a comment in script.

JoseAmador95 commented on bladeRF_MIMO_XCVR.m:
```diff
@@ -616,7 +648,8 @@
                     obj.vga2 = 0;
-                    obj.lna = 'MAX';
+                    %obj.lna = 'MAX'; % Trying to fix RX gain assignment
+                    obj.lna ='BYPASS'; % Original assignment
```
What is the purpose of this change?

danielwestofficial replied:
I will revert this to MAX, but left a comment on newest commit. I will check to ensure RX gain can be adjusted manually with obj.lna set to MAX.

danielwestofficial replied:
I just checked and setting obj.lna has no affect on setting the RX gain. Uncommented and commit.

JoseAmador95 commented on bladeRF_TestScript:
```diff
@@ -0,0 +1,67 @@
+%BW = Fs/2; % Lowering BW fulfills Nyquist Shannon Theorem
+fo = (frequencyStart + frequencyEnd)/2; % Center Frequency
+
+try 
```
Would it be possible to reduce the scope of the `try/catch` block so that errors can be reported with more detail?

JoseAmador95 commented on bladeRF_LoopbackTest:
```diff
@@ -0,0 +1,242 @@
+st = A * exp(1i * (w*tTXchirp + thetaTX + phi)); % Complex chirp
+
+
+%% Step 3: Initialize bladeRF Device for 2 TX and 2 RX Channels (MIMO)
```
This step duplicates the contents of _bladeRF_TestScript_. Would it be possible to reference such script instead of duplicating the contents in this file?

JoseAmador95 commented on bladeRF_LoopbackTest:
```diff
@@ -0,0 +1,242 @@
+%% Run
+
+try
+    for i = 1:64
```
Why 64?

danielwestofficial replied:
I observed that each iteration has a different time (the first 10 iterations tend to be very unstable). I am currently developing more for the loopback test that allows the shortest iteration to be used.

danielwestofficial replied:
I am currently working on enabling metadata in the TX and RX functions. My goal is to be able to accurately measure delay time between the TX and RX channels during a loopback test. I would expect something in nanoseconds range but am unable to get those results (I am getting 500ms). Thats where this iteration is being used for. This may be off topic, Ill send you info on Element and see how we want to proceed with this bladeRF_LoopbackTest.m script.

JoseAmador95 replied:
If 64 is an arbitrary number, that's ok. Just put it in a variable and add a comment explaining the value.

danielwestofficial replied:
It's somewhat arbitrary. I noticed on my setup that I was able to perform 64 iterations when receiving 1024 samples.

JoseAmador95 commented on bladeRF_LoopbackTest:
```diff
@@ -0,0 +1,242 @@
+device.rx.stop();
+
+% Save Receive Signal as a SC16 Q11 File
+%save_sc16q11('C:\Users\danie\OneDrive\Documents\OIT(FALL.2024)\ENGR596\Week10\HardwareInitialize(bladeRF_MIMOrev1)\tmp\sri.sc16q11', sri)
```
This path would only work on your PC. Can this be changed into a relative path?
```suggestion
save_sc16q11('tmp\sri.sc16q11', sri)
```

danielwestofficial replied:
Deleted.

JoseAmador95 commented on bladeRF_LoopbackTest:
Can we add an extension to this file?

danielwestofficial replied:
My thought process during the Fall term was to save the data to a file and the do post processing for the calculations. I ended leaving this code snippet in but it serves no purpose so can be removed.

danielwestofficial replied:
Deleted

JoseAmador95 commented on bladeRF_LoopbackTest:
There is some commented code. It should be either uncommented or deleted.

danielwestofficial replied:
Deleted

JoseAmador95 commented on bladeRF_LoopbackTest:
```diff
@@ -0,0 +1,242 @@
+%% Step 5: Verify and Plot Received Data
+
+% Load the samples
+%sri = load_sc16q11('sri.sc16q11');
```
I guess we don't need to load the data if it was just saved in the previous step

danielwestofficial replied:
Deleted

JoseAmador95 reviewed (changes requested):
The PR looks good overall. Let's just do some clean-up before merging. 👍🏽

danielwestofficial committed:
Update bladeRF_MIMO_XCVR.m

Removed chunk between lines 297-318
Reverted default parameters 
Deleted commented line on 631
Reverted lines 651-652 with comment (Will check that obj.lna = MAX can allow for manually setting RX gain).

danielwestofficial committed:
Update bladeRF_MIMO_XCVR.m

Uncommented line 625

danielwestofficial committed:
Update bladeRF_LoopbackTest

Deleted commented code.
